### PR TITLE
[Yaml] save preg_match() calls when possible

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | possibly #27960
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
This is how it looks like:

Parsing the [`services.yaml` file](https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/config/services.yaml) from the FrameworkBundle recipe:

bildschirmfoto 2018-08-01 um 20 45 06

(see 
https://blackfire.io/profiles/compare/b59c7350-088d-4a74-be43-45a39152ce89/graph)

Executing the Yaml component test suite:

bildschirmfoto 2018-08-01 um 20 49 28

(see https://blackfire.io/profiles/compare/b2f2e6a6-87d3-40ad-95dd-52a2ba6bdda7/graph)

**Review by fabpot — APPROVED** [review]

**Comment by staabm** [comment]
@xabbuh why does it save so much time in IOWait? The changes are not IO related..

File: src/Symfony/Component/Yaml/Parser.php [hunk 1]
```diff
@@ -1009,7 +1011,7 @@ private function isStringUnIndentedCollectionItem(): bool
      */
     private function isBlockScalarHeader(): bool
     {
-        return (bool) self::preg_match('~'.self::BLOCK_SCALAR_HEADER_PATTERN.'$~', $this->currentLine);
+        return '' !== $this->currentLine && (bool) self::preg_match('~'.self::BLOCK_SCALAR_HEADER_PATTERN.'$~', $this->currentLine);
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: what about adding `&& \in_array($this->currentLine[0], array('|', '>'), true) &&` here too ?

**Comment by stof** [comment]
@staabm probably cold vs warm OPCache cache (and so bypassing the IO for many files)

File: src/Symfony/Component/Yaml/Parser.php [hunk 2]
```diff
@@ -161,13 +161,13 @@ private function doParse(string $value, int $flags)
...
             $isRef = $mergeNode = false;
-            if (self::preg_match('#^\-((?P<leadspaces>\s+)(?P<value>.+))?$#u', rtrim($this->currentLine), $values)) {
+            if ('-' === $this->currentLine[0] && self::preg_match('#^\-((?P<leadspaces>\s+)(?P<value>.+))?$#u', rtrim($this->currentLine), $values)) {
                 if ($context && 'mapping' == $context) {
                     throw new ParseException('You cannot define a sequence item when in a mapping', $this->getRealCurrentLineNb() + 1, $this->currentLine, $this->filename);
                 }
                 $context = 'sequence';
 
-                if (isset($values['value']) && self::preg_match('#^&(?P<ref>[^ ]+) *(?P<value>.*)#u', $values['value'], $matches)) {
+                if (isset($values['value']) && '&' === $values['value'][0] && self::preg_match('#^&(?P<ref>[^ ]+) *(?P<value>.*)#u', $values['value'], $matches)) {
```
[the remark this answers is not in the document]
Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: No, the regular expression used to get here is `#^\-((?P<leadspaces>\s+)(?P<value>.+))?$#u`.

File: src/Symfony/Component/Yaml/Parser.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: I thought about that too, but the regular expression should also match lines like `foo: |`, shouldn't it?

Comment (stof) [reply src/Symfony/Component/Yaml/Parser.php]: hmm indeed. This regex is not anchored at the start. I missed that

File: src/Symfony/Component/Yaml/Parser.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (stof) [reply src/Symfony/Component/Yaml/Parser.php]: yeah, the `value` group can indeed not be empty when it is matched, so things are fine.

**Comment by xabbuh** [comment]
@staabm I think what @stof said. You can see updated profile comparisons at https://blackfire.io/profiles/compare/9de46b73-d563-4e0e-a81a-553cc3d5f290/graph (parsing the `services.yaml` file) and https://blackfire.io/profiles/compare/6effa0a7-6309-465e-a9a5-fb1863b96813/graph (running the Yaml component test suite).

File: src/Symfony/Component/Yaml/Parser.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: I just remembered that I indeed tried to add that check and was wondering now why tests didn't fail then. Turns out this method is totally useless with the current code (see #28187).

**Pull request merged**